src/cpp/server/server_cc.cc

```diff
@@ -410,11 +415,27 @@ class Server::CallbackRequest final : public internal::CompletionQueueTag {
       GPR_ASSERT(ignored == req_);
 
       if (!ok) {
-        // The call has been shutdown
-        req_->Clear();
+        // The call has been shutdown. Let it stay in the list
+        // since the server shutdown will delete and Clear it.
+        req_->Done();
         return;
       }
 
+      bool may_spawn_new = false;
+      {
+        std::lock_guard<std::mutex> l(req_->req_list_->reqs_mu);
+        req_->req_list_->reqs_list.erase(req_->req_list_iterator_);
+        // If this was the last request in the list, set up a new one
+        if (req_->req_list_->reqs_list.empty()) {
```

Comment:
For my own knowledge, is this when a new RPC starts in the server, and we are picking a pre-registered request from the list?

Reply:
Yes!

Reply:
Yes, I think that the sync and callback API could benefit from some changes to the request matcher (such as not having a request matcher at all and always just dynamically generating). I also suspect that doing that will take some time but is likely worth it.

Reply:
Filed #17754